internal/session/session.go

```diff
@@ -185,74 +184,97 @@ func (s *Session) Shutdown() {
 	s.shutdown()
 }
 
-// ReceiveFrom receives incoming blocks from the given peer.
-func (s *Session) ReceiveFrom(from peer.ID, ks []cid.Cid, haves []cid.Cid, dontHaves []cid.Cid) {
-	// The SessionManager tells each Session about all keys that it may be
-	// interested in. Here the Session filters the keys to the ones that this
-	// particular Session is interested in.
-	interestedRes := s.sim.FilterSessionInterested(s.id, ks, haves, dontHaves)
-	ks = interestedRes[0]
-	haves = interestedRes[1]
-	dontHaves = interestedRes[2]
-	s.logReceiveFrom(from, ks, haves, dontHaves)
+// GetBlock fetches a single block.
+func (s *Session) GetBlock(reqctx context.Context, k cid.Cid) (blocks.Block, error) {
+	if !k.Defined() {
+		log.Error("undefined cid in GetBlock")
+		return nil, blockstore.ErrNotFound
+	}
 
-	// Inform the session want sender that a message has been received
-	s.sws.Update(from, ks, haves, dontHaves)
+	gbctx, cancel := context.WithCancel(reqctx)
+	defer cancel()
 
-	if len(ks) == 0 {
-		return
+	// Call GetBlocks with the single key as an array
+	blkch, err := s.GetBlocks(gbctx, []cid.Cid{k})
+	if err != nil {
+		return nil, err
 	}
 
-	// Inform the session that blocks have been received
+	// Receive a single block from the channel
 	select {
-	case s.incoming <- op{op: opReceive, keys: ks}:
-	case <-s.ctx.Done():
+	case blk, ok := <-blkch:
+		if ok {
+			return blk, nil
+		}
+
+		// The channel should only be closed if the request context was
+		// cancelled
+		select {
+		case <-gbctx.Done():
+			return nil, gbctx.Err()
+		default:
+			return nil, errors.New("promise channel was closed")
+		}
+	case <-reqctx.Done():
+		return nil, reqctx.Err()
 	}
+
 }
 
-func (s *Session) logReceiveFrom(from peer.ID, interestedKs []cid.Cid, haves []cid.Cid, dontHaves []cid.Cid) {
-	// Save some CPU cycles if log level is higher than debug
-	if ce := sflog.Check(zap.DebugLevel, "Bitswap <- rcv message"); ce == nil {
-		return
+// GetBlocks fetches a set of blocks within the context of this session and
+// returns a channel that found blocks will be returned on. No order is
+// guaranteed on the returned blocks.
+func (s *Session) GetBlocks(ctx context.Context, keys []cid.Cid) (<-chan blocks.Block, error) {
+	// If there are no keys supplied, just return a closed channel
+	if len(keys) == 0 {
+		out := make(chan blocks.Block)
+		close(out)
+		return out, nil
 	}
 
-	for _, c := range interestedKs {
-		log.Debugw("Bitswap <- block", "local", s.self, "from", from, "cid", c, "session", s.id)
+	// Use a WantRequest to listen for incoming messages pertaining to the keys
+	wr, err := s.wrm.NewWantRequest(keys)
```

Comment:
How will this work when we switch to streaming requests?

Reply:
Currently `WantRequest.Run()` keeps track of which keys have been received, and closes the outgoing channel of blocks when all keys have been received.
I imagine that if we have a channel of incoming keys (instead of a slice) `WantRequest.Run()` will instead check if the channel is closed, and if so, close the outgoing channel of blocks.